File: examples/ruby/p384-primitives/P384Primitives.runar.rb
require 'runar'

class P384Primitives < Runar::SmartContract
  prop :expected_point, P384Point

  def initialize(expected_point)
    super(expected_point)
    @expected_point = expected_point
  end

  runar_public k: Bigint, base_point: P384Point
  def verify(k, base_point)
    result = p384_mul(base_point, k)
    assert p384_on_curve(result)
    assert result == @expected_point
  end

  runar_public a: P384Point, b: P384Point
  def verify_add(a, b)
    result = p384_add(a, b)
    assert p384_on_curve(result)
    assert result == @expected_point
  end

  runar_public k: Bigint
  def verify_mul_gen(k)
    result = p384_mul_gen(k)
    assert p384_on_curve(result)
    assert result == @expected_point
  end
end

File: examples/python/p384-primitives/P384Primitives.runar.py
from runar import SmartContract, ByteString, Bigint, public, assert_, p384_mul, p384_add, p384_mul_gen, p384_on_curve

class P384Primitives(SmartContract):
    expected_point: P384Point

    def __init__(self, expected_point: P384Point):
        super().__init__(expected_point)
        self.expected_point = expected_point

    @public
    def verify(self, k: Bigint, base_point: P384Point):
        result = p384_mul(base_point, k)
        assert_(p384_on_curve(result))
        assert_(result == self.expected_point)

    @public
    def verify_add(self, a: P384Point, b: P384Point):
        result = p384_add(a, b)
        assert_(p384_on_curve(result))
        assert_(result == self.expected_point)

    @public
    def verify_mul_gen(self, k: Bigint):
        result = p384_mul_gen(k)
        assert_(p384_on_curve(result))
        assert_(result == self.expected_point)
